### ISSUE TITLE
Allow stateful recurrent networks to be saved and loaded

### ISSUE THREAD

**Issue by hft7h11** [fchollet/keras#1363]
The code below breaks for stateful networks. I have tried manually editing the JSON to add in the batch_input_shape but that hasn't worked. 

save -
json_string = model.to_json()
open(fullNetFileName+'.json', 'w').write(json_string)
model.save_weights(fullNetFileName+'.h5')

load -
model = model_from_json(open(modelArch).read())
model.load_weights(modelWeights)

However now I get the following error: 
Exception: If a RNN is stateful, a complete input_shape must be provided (including batch size).

**Comment by hft7h11** [comment]
adding "batch_input_shape": [10,1, 5] after "input_shape" : [1,5] directly to the json doesn't seem to fix the issue. Same error. 

Replacing input_shape entirely with batch_input also doesn't fix the issue. The bug may also be in the inability of container_from_config to load the batch_input_shape property.

**Comment by hft7h11** [comment]
I believe I have identified why recurrent networks can't be loaded even after I added the batch_input_shape to the json

core.py lines 46 and 47

if 'batch_input_shape' in kwargs:
self.set_input_shape(tuple(kwargs['batch_input_shape']))

If you look at set_input_shape, it only ever sets the input_shape. So it looks like it is trying to set the input_shape to the batch_input_shape.

Also if you look at the two lines above

if 'input_shape' in kwargs and 'batch_input_shape' not in kwargs:
self.set_input_shape((None,) + tuple(kwargs['input_shape']))

It will already have called set_input_shape by then, which means it will call self.build(), which will break due to no batch_input_shape

This doesn't explain why the batch_input_shape isn't being saved to the json in the first place, but if I can at least get it to work after manually adding it in I will be happy. I am playing around with that now.
